test/e2e/node_feature_discovery_test.go

```diff
@@ -772,6 +772,128 @@ core:
+				var err error
+				crds, err = testutils.CreateNfdCRDs(extClient)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			AfterAll(func() {
+				for _, crd := range crds {
+					err := extClient.ApiextensionsV1().CustomResourceDefinitions().Delete(context.TODO(), crd.Name, metav1.DeleteOptions{})
+					Expect(err).NotTo(HaveOccurred())
+				}
+			})
+
+			BeforeEach(func() {
+				// Drop the pod security admission label as nfd-worker needs host mounts
+				if _, ok := f.Namespace.Labels[admissionapi.EnforceLevelLabel]; ok {
+					e2elog.Logf("Deleting %s label from the test namespace", admissionapi.EnforceLevelLabel)
+					delete(f.Namespace.Labels, admissionapi.EnforceLevelLabel)
+					_, err := f.ClientSet.CoreV1().Namespaces().Update(context.TODO(), f.Namespace, metav1.UpdateOptions{})
+					Expect(err).NotTo(HaveOccurred())
+				}
+				err := testutils.ConfigureRBAC(f.ClientSet, f.Namespace.Name)
+				Expect(err).NotTo(HaveOccurred())
+
+				// Remove pre-existing stale annotations and labels etc and CRDs
+				cleanupNode(f.ClientSet)
+				cleanupCRs(nfdClient, f.Namespace.Name)
+
+				By("Creating nfd-master config")
+				cm := testutils.NewConfigMap(configMapName, configMapKey, `
+denyLabelNs: ["*.denied.ns","random.unwanted.ns"]
+`)
+				cm, err = f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Create(context.TODO(), cm, metav1.CreateOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				// Launch nfd-master
+				By("Creating nfd-master pod")
+				podSpecOpts := createPodSpecOpts(
+					testpod.SpecWithContainerImage(dockerImage()),
+					testpod.SpecWithConfigMap(cm.Name, "/etc/kubernetes/node-feature-discovery"),
+					testpod.SpecWithTolerations(testTolerations),
+				)
+				masterPod := e2epod.NewPodClient(f).CreateSync(testpod.NFDMaster(podSpecOpts...))
+
+				// Create nfd-master service
+				nfdSvc, err := testutils.CreateService(f.ClientSet, f.Namespace.Name)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Waiting for the nfd-master pod to be running")
+				Expect(e2epod.WaitTimeoutForPodRunningInNamespace(f.ClientSet, masterPod.Name, masterPod.Namespace, time.Minute)).NotTo(HaveOccurred())
+
+				By("Verifying the node where nfd-master is running")
+				// Get updated masterPod object (we want to know where it was scheduled)
+				masterPod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), masterPod.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				// Node running nfd-master should have master version annotation
+				masterPodNode, err := f.ClientSet.CoreV1().Nodes().Get(context.TODO(), masterPod.Spec.NodeName, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				Expect(masterPodNode.Annotations).To(HaveKey(nfdv1alpha1.AnnotationNs + "/master.version"))
+
+				By("Waiting for the nfd-master service to be up")
+				Expect(e2enetwork.WaitForService(f.ClientSet, f.Namespace.Name, nfdSvc.Name, true, time.Second, 10*time.Second)).NotTo(HaveOccurred())
```

Comment:
We shouldn't repeat all of this setup teardown stuff. What you could do is e.g. something similar to what topologyupdater tests (`test/e2e/topology_updater_test.go`) do, combining `JustBeforeEach` and `BeforeEach`. Basically use `BeforeEach` in each test to set the exact parameters of nfd-master.